### DISCUSSION TITLE
New CLI Interface

### DISCUSSION THREAD

**Discussion by WoLfulus** [directus/directus#3084] [Feature Requests]
A new CLI interface would be awesome to automate some stuff around Directus. We already have some initial basic commands, but I'd love to get an overview of what we can do with it.

I'll keep the main post updated with the ongoing discussions/decisions

# Features/questions to ask

- [ ] Are we relying on REST/GraphQL to run the CLI? Standalone CLI?
  - [ ] Authentication
  - [ ] Context switches (multiple endpoints)
- [ ] Manage system collections
- [ ] Manage user collections
  - [ ] How to put those in the CLI? Which level?
- [ ] Supports the same API parameters (limit, sort, filter, offset, pagination, meta, search and export)
- [ ] Accepts JSON/YAML/RAW as input/output
  - [ ] Accept through `--arg` or `--stdin` to allow for piping `cat article.json | directus item create article`
  - [ ] Supports aggregations
- [ ] Supports proper casting to set values like `null`, `true` and `false`
- [ ] Doesn't clash fields with command arguments
- [ ] Supports aggregation
- [ ] Supports GraphQL
  - [ ] Supports variables for mutations
- [ ] Which output is the default? Raw?

# Examples

## Authenticate

```
$ echo password | directus login mysite https://directus.mysite.com --username wolfulus --password-stdin
Login successful

$ directus login mydev http://localhost:3000 --username testuser --password password
WARNING: Use --password-stdin 
Login successful

$ directus use-context mysite
Context switched to mysite (https://directus.mysite.com)

$ directus whereami
context: mysite
url: https://directus.mysite.com

$ directus whoami
wolfulus

$ directus use-context mydev
Context switched to mydev (http://localhost:3000)

$ directus whereami
context: mydev
url: http://localhost:3000

$ directus whoami
testuser
```

### Execute a query

`directus gql <data|system> [--stdin] [query]`

### Execute a query/mutation in user collections

`directus gql data "query { articles { slug title } }"`

### Execute a query/mutation in system collections

`directus gql system "query { user { email } }"`

### Bigger queries/mutations

`cat query.graphql | directus gql data --stdin`

```
cat <<GQL | directus gql data --stdin
  query {
    articles {
      slug
      title
    }
  }
GQL
```
`

### Execute a mutation

```
cat <<GQL | directus gql data --stdin
  mutation {
    createArticle(article: {
      title: "Hello"
    }) {
      id
    }
  }
GQL
```

### Create a system item

> These can be added at the root level

`directus user create --email <value>`
`directus user create --email=<value>`

### Create a data item

> Scoped under `items`

`directus items create [create_options] <collection> --<column>=?<value>`

## Fetch items

`directus items get [options] <collection> [fields]`

### Aggregation

`directus items get --count articles`

> How

### Don't return data

`directus items get --no-data articles` 

## Update items

> TBD

## Delete items

> TBD

## Input methods

```
$ directus --yaml items create article --title="My title" --slug="my-title"
data: 
  id: 5
  title: "My title"
  slug: "my-title"
  content: ""
  create_at: "..."
  updated_at: ~

$ cat article.json | directus --json items create --stdin article
{
  "data": {
    "id": 5,
    "title": "My title",
    "slug": "my-title",
    "content": "",
    "created_at": "...",
    "updated_at": null
  }
}
```

## Output control

```
$ directus --yaml <command>
data:
  - value: 1234

$ directus --json <command>
{
  "data": {
    "value": 1234
  }
}

$ directus --dot <command>
data.value=1234

$ directus --raw <command>
????

$ directus <command>
????

```

**Comment by rijkvanzanten** [comment]
I think the external API management is powerful and useful, but a tad bit unexpected seeing the CLI is shipped with the particular install of Directus. I think we should focus our efforts on management of the local DB of the current installation.

Having a standard structure for IO for all things (system + items) is important, I think "matching" the REST API makes sense here, for example:

```
POST /users
directus users create
```

Similarly, it would be useful if it would use the same IO as the endpoints, as that means we can use the exact same internals and don't have to do any conversion magic:

```
POST /users

{
  "email": "admin@example.com",
  "password": "password"
}

--

directus users create "{ \"email\": \"admin@example.com\", \"password\": \"password\" }"
```

While the syntax is a little nasty to read, it would allow us to use args for the "query params": 

`directus users create "{...}" --fields id,email,avatar.*`

We could then also "default" to `stdin` if the object itself is left empty, which would allow for @WoLfulus's example above:

```
cat article.json | directus items articles create --fields *.*.* --meta total_count
```

**Comment by rijkvanzanten** [comment]
Re command, we could go with

```
directus users create
```

_or_

have it be create, and rely on the `directus_` prefix for the system stuff:

```
directus create <collection>
```

Reads a little nicer in use:

```
directus read directus_users 
```

```
directus read articles
```

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
